### PR TITLE
Feature/ado 2622 challenge wizard (Features 1, 2 & 3)

### PR DESCRIPTION
**Pull request by geoffroth**
### Description
This PR would add several new components related to challenges:
- `ChallengeRegisterButton` - A button that changes state depending on the user's status vis-a-vis the current challenge
- `ChallengeTeamWizard` - A series of modal screens that walk the user through joining a challenge, as per the [Figma spec](https://www.figma.com/file/Asjvjxzn1lGktlDye6WCB0/Challenge-Portal-with-GenUI?type=design&node-id=1-11376&t=dflE0DdyzkO12ttP-0)
- `ChallengeDetailPage` - A stub page that embed the ChallengeRegisterButton allowing for some basic manual testing of the overall flow
- A number of supporting components including `StepperDialog`, a component for creating next / previous step flows

The following services were added to SynapseClient in support of this:
- `getChallengeTeams` - lists Teams registered to the Challenge
- `getSubmissionTeams` - lists Teams under which the given submitter may submit to the Challenge
- `registerChallengeTeam` - registers a Team with a Challenge
- `createTeam` - creates a new Team
- `createMembershipInvitation` - creates a new membership invitation and sends an email to the invitee
- `getMembershipStatus` - retrieve the Team Membership Status bundle for a team and user
- `createMembershipRequest` - creates a membership request and sends an email to the administrators of the team.
- `deleteMemberFromTeam` - removes the given member from the specified Team
- `getTeamList` - gets Teams that match the given list of IDs

There are several new associated types and hooks corresponding to the above additions.

### CODE REVIEW

File: packages/synapse-react-client/package.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
     "@mui/material": "^5.11.16",
     "@mui/system": "^5.11.16",
     "@mui/utils": "^5.11.13",
+    "@mui/x-data-grid": "5.17.26",
```
Comment (geoffroth) [inline packages/synapse-react-client/package.json:40]: Added MUI Datagrid Component

File: packages/synapse-react-client/src/components/ChallengeDetailPage/ChallengeDetailPage.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+       * Should the team be removed from the challenge instead?
+       * Only a team admin can request this however.
+       */
+      /*
+      deleteMemberFromTeam(
+        challenge.participantTeamId,
+        userProfile.ownerId,
+        accessToken,
+      )
+      */
```
Comment (geoffroth) [inline packages/synapse-react-client/src/components/ChallengeDetailPage/ChallengeDetailPage.tsx:54]: Removing a member from the challenge participant team does not remove them from any submission teams they may belong to, and leaves the user in an ambiguous state vis-a-vis the challenge. Further discussion may be required to determine what steps are required here, or whether leaving should be offered as an option at all.

File: apps/portals/src/portal-components/elportal/ELBetaLaunchBanner.tsx [hunk 3]
```diff
@@ -15,7 +15,7 @@ const ELBetaLaunchBanner = () => {
       }}
       variant={'info'}
       show={true}
-      title='Welcome to the beta version of the ELITE Portal!'
+      title="Welcome to the beta version of the ELITE Portal!"
```
Comment (geoffroth) [inline apps/portals/src/portal-components/elportal/ELBetaLaunchBanner.tsx:18]: Apologies, a number of small cosmetic changes like this happened when i merged in main. I may be wrong but I think they match the defined linting settings, so I did not revert them.

**Review by nickgros — APPROVED** [review]
Generally looks good, there are a number of style issues that, given the timeline, need not block merging, which I'll track in new issues on our Jira board (pointing to my comments here as a reference). Besides that, there's one copy typo (which I'll fix in the PR UI).

There are two other issues I want to raise, but there's no reason those need to be addressed to merge this.

1. The ChallengeTeamList likely doesn't handle pagination, since I doubt the backend respects the requested 500 items limit. It's not clear to me if this is a requirement.
2. Our membership invitation service throttles very aggressively, which we may want to capture in the UI. Discussions are needed from the Sage team, so no action needed yet.

File: packages/synapse-react-client/stories/StepperDialog.tsx
[the remark this answers is not in the document]
Comment (geoffroth) [reply packages/synapse-react-client/stories/StepperDialog.tsx:1]: Story file was misnamed. Should be fixed now.

File: packages/synapse-react-client/src/components/ChallengeTeamWizard/ChallengeTeamTable.tsx [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+export default function ChallengeTeamTable({
+  challengeId,
+  onSelectTeam,
+}: ChallengeTeamTableProps) {
+  const [allRows, setAllRows] = useState<ChallengeTeamRow[]>([])
+  const [teamIdList, setTeamIdList] = useState<string[]>([])
+  const [teamsById, setTeamsById] = useState<Record<string, Team>>({})
+  const { data: regTeams } = useGetChallengeTeamList(challengeId, 0, 500)
```
[the remark this answers is not in the document]
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/ChallengeTeamWizard/ChallengeTeamTable.tsx:35]: @geoffroth : We should properly support pagination here.  Please see useEntity.ts useGetVersionsInfinite() definition, which customizes the next page param using a static limit (you could set to something like 20 challenge teams at a time).
Then take a look at AccessRequirementTable on the pattern to utilize the hasNextPage value to offer a "Show More" button.

File: packages/synapse-react-client/src/components/ChallengeTeamWizard/ChallengeTeamWizard.tsx [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+      const errors: string[] = []
+      for (const inviteeEmail of emails) {
+        await createMembershipInvitation(
+          accessToken,
+          teamId,
+          inviteeEmail.trim(),
+        ).catch(() => {
+          errors.push(inviteeEmail.trim())
+        })
+      }
```
[the remark this answers is not in the document]
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/ChallengeTeamWizard/ChallengeTeamWizard.tsx:464]: https://www.figma.com/file/Asjvjxzn1lGktlDye6WCB0?node-id=1:11777#467050802

[the remark this answers is not in the document]
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/ChallengeTeamWizard/ChallengeTeamWizard.tsx:464]: Added a comment in the design to update if user would like to invite more than 3 people to their new team

File: packages/synapse-react-client/src/components/ChallengeTeamWizard/CreateChallengeTeam.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        />
+      </Box>
+      <TextField
+        id="invitees"
+        label="Emails of Additional Members to Invite"
+        placeholder="Enter emails separated by comma"
+        value={team.invitees}
+        fullWidth
+        onChange={event => handleTeamUpdate({ invitees: event.target.value })}
+      />
```
[the remark this answers is not in the document]
Comment (geoffroth) [reply packages/synapse-react-client/src/components/ChallengeTeamWizard/CreateChallengeTeam.tsx]: Now limiting to 3 invitees on team creation

File: packages/synapse-react-client/src/components/ChallengeTeamWizard/SelectChallengeTeam.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+          onSelectTeam={onSelectTeam}
+        />
+      </Box>
+      <Box>
+        <Button
+          color="primary"
+          variant="contained"
+          onClick={onCreateTeam}
+          startIcon={<ControlPointIcon />}
+          // sx={{ marginTop: '50px', padding: '6px 18px' }}
```
[the remark this answers is not in the document]
Comment (geoffroth) [reply packages/synapse-react-client/src/components/ChallengeTeamWizard/SelectChallengeTeam.tsx:40]: Removed

File: packages/synapse-react-client/src/components/ChallengeTeamWizard/ChallengeTeamWizard.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (geoffroth) [reply packages/synapse-react-client/src/components/ChallengeTeamWizard/ChallengeTeamWizard.tsx:439]: Limiting to 3 invitees, and messaging if user attempts to add more.

File: packages/synapse-react-client/src/components/ChallengeTeamWizard/CreateChallengeTeam.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+  const handleTeamUpdate = (update: Record<string, string>) => {
+    const updatedTeam: CreateTeamRequest = { ...team, ...update }
+    setTeam(updatedTeam)
+    onChangeTeamInfo(updatedTeam)
+  }
+
+  return (
+    <Box>
+      <Box>Create a new team for this Challenge!</Box>
```
[the remark this answers is not in the document]
Comment (geoffroth) [reply packages/synapse-react-client/src/components/ChallengeTeamWizard/CreateChallengeTeam.tsx]: Replaced here and other similar locations

File: packages/synapse-react-client/src/synapse-client/SynapseClient.ts [hunk 9]
```diff
@@ -1383,6 +1389,79 @@ export const getUserChallenges = (
...
+ * see http://rest-docs.synapse.org/rest/GET/challenge.html
+ */
+export const getAllChallengeTeams = (
+  accessToken: string | undefined,
+  challengeId: string | number,
+  offset: string | number = 0,
+  limit: string | number = 200,
+): Promise<ChallengeTeam[]> => {
+  const url = `/repo/v1/challenge/${challengeId}/challengeTeam?&offset=${offset}&limit=${limit}`
+  const fn = () => {
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/synapse-client/SynapseClient.ts:1417]: This function should take the limit and offset for getAllOfPaginatedService() to work properly

(limit: number, offset: number) => {
    const url = `/repo/v1/challenge/${challengeId}/challengeTeam?&offset=${offset}&limit=${limit}`

Comment (jay-hodgson) [reply packages/synapse-react-client/src/synapse-client/SynapseClient.ts:1417]: Does not make sense for getAllChallengeTeams() to take an offset or limit

File: packages/synapse-react-client/src/synapse-client/SynapseClient.ts [hunk 10]
```diff
@@ -1383,6 +1389,79 @@ export const getUserChallenges = (
...
+/**
+ * List the Teams under which the given submitter may submit to the Challenge,
+ * i.e. the Teams on which the user is a member and which are registered for the Challenge.
+ * see https://rest-docs.synapse.org/rest/GET/challenge/challengeId/submissionTeams.html
+ */
+export const getSubmissionTeams = (
+  accessToken: string | undefined,
+  challengeId: string | number,
+  offset: string | number = 0,
+  limit: string | number = 200,
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/synapse-client/SynapseClient.ts:1436]: Again, I think the backend will ignore this limit when it considers it too high (usually supports 20-50).
This should also probably use the getAllOfPaginatedService()

Comment (geoffroth) [reply packages/synapse-react-client/src/synapse-client/SynapseClient.ts:1436]: The 200 was just me following a pattern I found sprinkled throughout existing calls in SynapseClient. Happy to reduce.

Comment (geoffroth) [reply packages/synapse-react-client/src/synapse-client/SynapseClient.ts:1436]: I'm only requesting 1 or 2 depending on the context.

File: packages/synapse-react-client/src/synapse-client/SynapseClient.ts (hunk 9 above)
Comment (geoffroth) [reply packages/synapse-react-client/src/synapse-client/SynapseClient.ts:1417]: Good catch

Comment (geoffroth) [reply packages/synapse-react-client/src/synapse-client/SynapseClient.ts]: I had unwittingly used another existing errant function as the basis for the structure of this one. Should be fixed now in both places.

File: packages/synapse-react-client/src/synapse-client/SynapseClient.ts (hunk 10 above)
Comment (geoffroth) [reply packages/synapse-react-client/src/synapse-client/SynapseClient.ts]: Reduced default to 50

**Pull request merged**